File: plusnine_custom/public/js/custom_lead.js
frappe.ui.form.on('Lead', {
    refresh: function(frm) {
        // frm.timeline.insert_comment('Comment', 'This is an automated comment.');
        setTimeout(() => {
            cur_frm.page.remove_inner_button(__('Customer'),  __('Create'));
       }, 500);
    },
    custom_vehicle_type: function(frm) {
        if (frm.doc.custom_vehicle_type === "Car") {
             frm.set_value('custom_bike_type', '');
             frm.set_value('custom_bike_brand', '');
             frm.set_value('custom_bike_model_name', '');
         } else {
             frm.set_value('custom_body_type', '');
             frm.set_value('custom_car_brand', '');
             frm.set_value('custom_car_model_name', '');
         }
     }
}); 


 
frappe.ui.form.on('Lead', {
    refresh: function(frm) {
        frm.add_custom_button(__("Add Comment"), function() {
            if(!frm.is_new()){  
                let d = new frappe.ui.Dialog({
                title: 'Enter Comment',
                fields: [
                    {
                        label: 'Comment', 
                        fieldname: 'comment',
                        fieldtype: 'Small Text',
                        reqd:1
                    }
                 
                ],
                size: 'small', 
                primary_action_label: 'Add Comment',
                async primary_action(values) {
                    frappe.call({
                        method:"plusnine_custom.public.py.custom_lead.add_comments",
                        args:{
                            comment:values.comment,
                            email: frm.doc.email_id,
                            name: frm.doc.name
                        },
                        callback:(response)=>{
                        if(response.message == "Success"){
                            frappe.msgprint("Comment Added Successfully")
                        }else{
                            frappe.msgprint("Failed to Add Comment")
                        }
                            
                        }
                    })
                    d.hide();
                }
            });
                    d.show();
               
            }
            
        });


    //   if (!frm.is_new()) {
    //         frm.add_custom_button(__('Add Event'), function () {

    //             let dialog = new frappe.ui.Dialog({
    //                 title: 'Add Event',
    //                 fields: [
    //                     {
    //                         fieldname: 'subject',
    //                         fieldtype: 'Data',
    //                         label: 'Subject',
    //                         reqd: 1,
    //                         default: `Meeting with ${frm.doc.lead_name}`
    //                     },
    //                     {
    //                         fieldname: 'starts_on',
    //                         fieldtype: 'Datetime',
    //                         label: 'Start Time',
    //                         reqd: 1,
    //                         default: frappe.datetime.now_datetime()
    //                     },
    //                     {
    //                         fieldname: 'duration',
    //                         fieldtype: 'Int',
    //                         label: 'Duration (Minutes)',
    //                         default: 30
    //                     },
    //                     {
    //                         fieldname: "Event Category",
    //                         fieldtype: "Select",
    //                         label: "Event Category",
    //                         options: [
    //                             "Event",
    //                             "Meeting",
    //                             "Call",
    //                             "Sent/Received Email",
    //                             "Other",
    //                             "Follow Up",
    //                             "Studio Visit",
    //                         ],
    //                         default: "Event"

    //                     },
    //                     {
    //                         fieldname: 'assign_users',
    //                         fieldtype: 'MultiSelectPills',
    //                         label: 'Assign To',
    //                         get_data: function (txt) {
    //                             return frappe.db.get_link_options('User', txt);
    //                         }
    //                     },
    //                     {
    //                         fieldname: 'description',
    //                         fieldtype: 'Small Text',
    //                         label: 'Description'
    //                     }
    //                 ],
    //                 primary_action_label: 'Save',
    //                 primary_action(values) {

    //                     let ends_on = moment(values.starts_on)
    //                         .add(values.duration || 30, 'minutes')
    //                         .format('YYYY-MM-DD HH:mm:ss');

    //                     frappe.call({
    //                         method: 'frappe.client.insert',
    //                         args: {
    //                             doc: {
    //                                 doctype: 'Event',
    //                                 subject: values.subject,
    //                                 starts_on: values.starts_on,
    //                                 ends_on: ends_on,
    //                                 description: values.description,
    //                                 event_participants: [{
    //                                     reference_doctype: 'Lead',
    //                                     reference_docname: frm.doc.name
    //                                 }]
    //                             }
    //                         },
    //                         callback: function () {
    //                             frappe.msgprint(__('Event created successfully'));
    //                             dialog.hide();
    //                         }
    //                     });
    //                 }
    //             });

    //             dialog.show();
    //         });
    //     }


 // ================= ADD EVENT =================
        if (frm.is_new()) return;
        frm.add_custom_button(__('Add Follow Up'), function () {

            let dialog = new frappe.ui.Dialog({
                title: 'Add Event',
                fields: [
                    {
                        fieldname: 'subject',
                        fieldtype: 'Data',
                        label: 'Subject',
                        reqd: 1,
                        default: `Meeting with ${frm.doc.lead_name}`
                    },
                    {
                        fieldname: 'starts_on',
                        fieldtype: 'Datetime',
                        label: 'Start Time',
                        reqd: 1,
                        default: frappe.datetime.now_datetime()
                    },
                    {
                        fieldname: 'ends_on',
                        fieldtype: 'Datetime',
                        label: 'End Time',
                        reqd: 1,
                        default: frappe.datetime.now_datetime()
                    },
                    {
                        fieldname: 'duration',
                        fieldtype: 'Int',
                        label: 'Duration (Minutes)',
                        default: 30
                    },
                    {
                        fieldname: 'event_category',
                        fieldtype: 'Select',
                        label: 'Event Category',
                        options: [
                            'Event',
                            'Meeting',
                            'Call',
                            'Sent/Received Email',
                            'Other',
                            'Follow Up',
                            'Studio Visit'
                        ],
                        default: 'Event'
                    },
                    {
                        fieldname: 'assign_users',
                        fieldtype: 'MultiSelectPills',
                        label: 'Assign To',
                        get_data: function (txt) {
                            return frappe.db.get_link_options('User', txt);
                        }
                    },
                    {
                        fieldname: 'description',
                        fieldtype: 'Small Text',
                        label: 'Description'
                    }
                ],
                primary_action_label: 'Save',
                primary_action(values) {

                    frappe.call({
                        method: "plusnine_custom.public.py.custom_lead.create_event_with_todos",
                        args: {
                            data: {
                                subject: values.subject,
                                starts_on: values.starts_on,
                                ends_on: values.ends_on,
                                duration: values.duration,
                                description: values.description,
                                event_category: values.event_category,
                                lead_name: frm.doc.name,
                                assign_users: values.assign_users || []
                            }
                        },
                        callback: function () {
                            frappe.msgprint(__('Event and ToDo(s) created successfully'));
                            dialog.hide();
                        }
                    });
                }
            });

            dialog.show();
        });



        if (frm.is_new()) return;

        frm.add_custom_button(__('RNR Follow Up'), function () {

            let dialog = new frappe.ui.Dialog({
                title: 'Add Event',
                fields: [
                    {
                        fieldname: 'subject',
                        fieldtype: 'Data',
                        label: 'Subject',
                        reqd: 1,
                        default: `Fallow Up Call`
                    },
                    {
                        fieldname: 'starts_on',
                        fieldtype: 'Datetime',
                        label: 'Start Time',
                        reqd: 1,
                        default: frappe.datetime.now_datetime()
                    },
                    {
                        fieldname: 'event_category',
                        fieldtype: 'Select',
                        label: 'Event Category',
                        options: [
                            'Event',
                            'Meeting',
                            'Call',
                            'Sent/Received Email',
                            'Other',
                            'Follow Up',
                            'Studio Visit'
                        ],
                        default: 'Follow Up'
                    },
                    // {
                    //     fieldname: 'assign_users',
                    //     fieldtype: 'MultiSelectPills',
                    //     label: 'Assign To',
                    //     get_data: function (txt) {
                    //         return frappe.db.get_link_options('User', txt);
                    //     }
                    // },
                    {
                        fieldname: 'description',
                        fieldtype: 'Small Text',
                        label: 'Description'
                    }
                ],
                primary_action_label: 'Save',
                primary_action(values) {

                    frappe.call({
                        method: "plusnine_custom.public.py.custom_lead.create_event_with_todos_rnr",
                        args: {
                            data: {
                                subject: values.subject,
                                starts_on: values.starts_on,
                                description: values.description,
                                event_category: values.event_category,
                                lead_name: frm.doc.name,
                                assign_users: values.assign_users || [],
                                created_from_button: 1  
                            }
                        },
                        callback: function () {
                            frappe.msgprint(__('Event and ToDo(s) created successfully'));
                            dialog.hide();
                        }
                    });
                }
            });

            dialog.show();
        });



    }
}); 



// frappe.ui.form.on('Lead', {
//     refresh: function(frm) {
//         // Ensure the form is fully loaded before inserting a comment
//         setTimeout(() => {
//             frappe.model.with_doc("Lead", frm.doc.name, function() {
//                 frappe.model.add_comment("Lead", frm.doc.name, "This is an automated comment.")
//                     .then(() => {
//                         frm.reload_doc(); // Reload to show the comment
//                     });
//             });
//         }, 500);

//         // Remove only the specific "Customer" button, ensuring others remain
//         setTimeout(() => {
//             frm.page.wrapper.find('.dropdown-menu .dropdown-item').each(function() {
//                 if ($(this).text().trim() === __('Customer')) {
//                     $(this).remove();
//                 }
//             });
//         }, 500);
//     },

//     custom_vehicle_type: function(frm) {
//         if (frm.doc.custom_vehicle_type === "Car") {
//             frm.set_value('custom_bike_type', '');
//             frm.set_value('custom_bike_brand', '');
//             frm.set_value('custom_bike_model_name', '');
//         } else {
//             frm.set_value('custom_body_type', '');
//             frm.set_value('custom_car_brand', '');
//             frm.set_value('custom_car_model_name', '');
//         }
//     }
// });

File: plusnine_custom/public/py/custom_lead.py
import frappe
import re
from frappe.utils import now_datetime, add_to_date
# import time
  
@frappe.whitelist()  
def assign_sales_partner(doc,method): 
     
    # time.sleep(10)

    # if not (doc.custom_pincode and doc.custom_brand and doc.type):
    #     return
 
    user_lists = frappe.get_all("Sales Partner", fields=["name"])
    matched_users = []

    for user in user_lists:
        user_doc = frappe.get_doc("Sales Partner", user.name)

    # //check is exists or not
    #   frappe.db.exists("table name",{"parent":user.name,"pincodes":doc.custom_pincode})
          
        has_matching_pincode = any(str(row.pincodes) == str(doc.custom_pincode) for row in user_doc.custom_pincode)
        has_matching_brand = any(str(row.brand) == str(doc.custom_brand) for row in user_doc.custom_brand)
        has_matching_type = any(str(row_lead.lead_type) == str(doc.type) for row_lead in user_doc.custom_lead_type)
  
        if has_matching_pincode and has_matching_brand and has_matching_type:
            matched_users.append(user.name)

    if not matched_users:
        frappe.msgprint(f"No matching Sales Partner found for:\nPincode: {doc.custom_pincode}, Brand: {doc.custom_brand}, Lead Type: {doc.type}")
        # doc.custom_lead_status = "Not Assigned" 
        frappe.db.set_value("Lead", doc.name, "custom_lead_status", "Not Assigned")
        
        if doc.custom_sales_partner:
            matched_users = [doc.custom_sales_partner]

    for sales_partner_name in matched_users:
        customer = frappe.get_doc('Sales Partner', sales_partner_name)

        # doc.custom_lead_status =  "Assigned"
        frappe.db.set_value("Lead", doc.name, "custom_lead_status", "Assigned")


        if customer.custom_customer:
            # frappe.db.set_value("Lead", doc.name, "custom_sales_partner_customer", customer.custom_customer)
            doc.custom_sales_partner_customer =  customer.custom_customer
            customer_doc = frappe.get_doc("Customer", customer.custom_customer)
            if customer_doc.email_id:
                # frappe.db.set_value("Lead", doc.name, "custom_customer_email_id", customer_doc.email_id)
                doc.custom_customer_email_id = customer_doc.email_id

        if customer.custom_email:
            # frappe.db.set_value("Lead", doc.name, "custom_sales_partner_email", customer.custom_email)
            doc.custom_sales_partner_email = customer.custom_email

        # Avoid duplicate assignment
        exists = frappe.db.exists("Sales Partner Assigned Lead", {
            "document_type": "Lead",
            "document_id": doc.name,
            # "sales_partner": sales_partner_name
        })
        if exists:
            #  assigned_doc = frappe.get_doc("Sales Partner Assigned Lead", exists)
            #  frappe.throw(str(assigned_doc.sales_partner))     
            assigned_doc = frappe.get_doc("Sales Partner Assigned Lead", exists)

            # Update the fields using the new Lead (doc) info
            assigned_doc.update({
                "document_type": "Lead",
                "lead_type": doc.type,
                "document_id": doc.name,
                "sales_partner": sales_partner_name,
                "lead_name": doc.first_name,
                "lead_email": doc.email_id,
                "lead_city": doc.city,
                "lead_pincode": doc.custom_pincode,
                "lead_status": doc.status,
                "lead_brand": doc.custom_brand,
                "sales_partner_mobile_number" : customer.custom_mobile_no,
                "sales_partner_customer" : customer.custom_customer,
                "lead_mobile_no" : doc.mobile_no,
            })

            assigned_doc.save(ignore_permissions=True)
            frappe.msgprint(f"Updated existing Sales Partner Assigned Lead: {sales_partner_name}")

        else: 
            comments = frappe.get_all("Comment", filters={
                "reference_doctype": "Lead",
                "reference_name": doc.name,  
                "comment_type": "Comment"
            },limit_page_length = 10, fields=["content"])

            comment_text = "\n".join(
                re.sub(r'<[^>]+>', '', c["content"]).strip()
                for c in comments if c.get("content")  
            )
       
            new_doc = frappe.new_doc('Sales Partner Assigned Lead')
            new_doc.document_type = "Lead"
            new_doc.document_id = doc.name 
            new_doc.sales_partner = sales_partner_name
            new_doc.lead_type = doc.type
            new_doc.lead_name = doc.first_name
            new_doc.lead_email = doc.email_id
            new_doc.lead_city = doc.city
            new_doc.lead_pincode = doc.custom_pincode
            new_doc.lead_status = doc.status
            new_doc.lead_brand = doc.custom_brand
            new_doc.sales_partner_mobile_number = customer.custom_mobile_no
            new_doc.sales_partner_customer = customer.custom_customer
            new_doc.lead_mobile_no = doc.mobile_no
            new_doc.lead_comments = comment_text


            # new_doc.update({ 
            #     "document_type": "Lead",
            #     "document_id": doc.name,
            #     "sales_partner": sales_partner_name,
            #     "lead_type": doc.type,
            #     "lead_name": doc.first_name,
            #     "lead_email": doc.email_id,
            #     "lead_city": doc.city,
            #     "lead_pincode": doc.custom_pincode,
            #     "lead_status": doc.status,
            #     "lead_brand": doc.custom_brand,
            #     "sales_partner_mobile_number": customer.custom_mobile_no,
            #     "sales_partner_customer": customer.custom_customer,
            #     "lead_mobile_no": doc.mobile_no,
            #     "lead_comments": comment_text
            # })
            # frappe.msgprint(str(new_doc))
            new_doc.save(ignore_permissions=True)


 
@frappe.whitelist()
def add_comments(comment,name, email=None):
    comment_doc = frappe.new_doc("Comment")
    comment_doc.reference_doctype = "Lead"
    comment_doc.reference_name = name
    comment_doc.comment_email = email
    comment_doc.comment_type = "Comment"
    comment_doc.content = comment
    comment_doc.save()
    return "Success"





# @frappe.whitelist()
# def create_event_with_todos(data):
#     """
#     data = {
#         subject,
#         starts_on,
#         ends_on,
#         duration,
#         description,
#         event_category,
#         lead_name,
#         assign_users: []
#     }
#     """

#     data = frappe.parse_json(data)

#     users = [u["value"] for u in data.get("assign_users", [])]

#     # 1️⃣ Create Event
#     event = frappe.get_doc({
#         "doctype": "Event",
#         "subject": data["subject"],
#         "starts_on": data["starts_on"],
#         "ends_on": data["ends_on"],
#         "event_category": data.get("event_category"),
#         "description": data.get("description"),
#         "event_participants": [{
#             "reference_doctype": "Lead",
#             "reference_docname": data["lead_name"]
#         }],
#         "custom_assign_to": [
#             {"user": user} for user in users
#         ]
#     })

#     event.insert(ignore_permissions=True)

#     # 2️⃣ Create / Update ToDos (SEQUENTIAL & SAFE)
#     for user in data.get("assign_users", []):

#         todo_name = frappe.db.get_value(
#             "ToDo",
#             {
#                 "reference_type": "Event",
#                 "reference_name": event.name,
#                 "allocated_to": user
#             },
#             "name"
#         )

#         if todo_name:
#             frappe.db.set_value(
#                 "ToDo",
#                 todo_name,
#                 {
#                     "description": data["subject"],
#                     "status": "Open",
#                     "date": data["starts_on"]
#                 }
#             )
#         else:
#             todo = frappe.get_doc({
#                 "doctype": "ToDo",
#                 "allocated_to": user,
#                 "reference_type": "Event",
#                 "reference_name": event.name,
#                 "description": data["subject"],
#                 "status": "Open",
#                 "date": data["starts_on"]
#             })
#             todo.insert(ignore_permissions=True)

#     frappe.db.commit()
#     return event.name


@frappe.whitelist()
def create_event_with_todos(data):

    data = frappe.parse_json(data)

    # ✅ Normalize assign_users (string list OR object list)
    users = []
    for u in data.get("assign_users", []):
        if isinstance(u, dict):
            users.append(u.get("value"))
        else:
            users.append(u)

    # 1️⃣ Create Event
    event = frappe.get_doc({
        "doctype": "Event",
        "subject": data["subject"],
        "starts_on": data["starts_on"],
        "ends_on": data["ends_on"],
        "event_category": data.get("event_category"),
        "description": data.get("description"),
        "event_participants": [{
            "reference_doctype": "Lead",
            "reference_docname": data["lead_name"]
        }],
        "custom_assign_to": [
            {"user": user} for user in users
        ]
    })

    event.insert(ignore_permissions=True)

    # 2️⃣ Create / Update ToDos (USE users, NOT raw data)
    for user in users:

        todo_name = frappe.db.get_value(
            "ToDo",
            {
                "reference_type": "Event",
                "reference_name": event.name,
                "allocated_to": user
            },
            "name"
        )

        if todo_name:
            frappe.db.set_value(
                "ToDo",
                todo_name,
                {
                    "description": data["subject"],
                    "status": "Open",
                    "date": data["starts_on"]
                }
            )
        else:
            frappe.get_doc({
                "doctype": "ToDo",
                "allocated_to": user,
                "reference_type": "Event",
                "reference_name": event.name,
                "description": data["subject"],
                "status": "Open",
                "date": data["starts_on"]
            }).insert(ignore_permissions=True)

    frappe.db.commit()
    return event.name



import frappe
from frappe.utils import add_days, now_datetime

@frappe.whitelist()
def create_event_with_todos_rnr(data):

    data = frappe.parse_json(data)
    lead_name = data.get("lead_name")

    # ✅ Always set start date = Today + 2 days
    starts_on = add_days(now_datetime(), 2)

    session_user = frappe.session.user

    # ✅ Create RNR Event
    event = frappe.get_doc({
        "doctype": "Event",
        "subject": "Followup Call",
        "starts_on": starts_on,
        "event_category": "Call",
        "custom_is_rnr": 1,
        "event_participants": [{
            "reference_doctype": "Lead",
            "reference_docname": lead_name
        }],
        "custom_assign_to": [
            {"user": session_user}
        ]
    })

    event.insert(ignore_permissions=True)

    # ✅ Create RNR ToDo
    frappe.get_doc({
        "doctype": "ToDo",
        "allocated_to": session_user,
        "reference_type": "Event",
        "reference_name": event.name,
        "description": "Followup Call",
        "status": "Open",
        "date": starts_on,
        "custom_is_rnr": 1
    }).insert(ignore_permissions=True)

    frappe.db.commit()
    return event.name
